nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/model/impl/DeprecatedProperty.java

```diff
@@ -0,0 +1,213 @@
+        } else {
+            log.warn(msg);
+        }
+        deprecated.setValue(value);
+        if (fallback != null) {
+            fallback.setValue(value);
+        }
+    }
+
+    @Override
+    public Serializable internalGetValue() throws PropertyException {
+        StringBuilder msg = newDeprecatedMessage();
+        if (fallback == null) {
+            msg.append("Return value from deprecated property");
+        } else {
+            msg.append("Return value from '")
+               .append(fallback.getPath())
+               .append("' if not null, from deprecated property otherwise");
+        }
+        if (log.isTraceEnabled()) {
+            log.warn(msg, new NuxeoException());
+        } else {
+            log.warn(msg);
+        }
+        Serializable deprecatedValue = deprecated.getValue();
+        if (fallback == null) {
+            return deprecatedValue;
+        }
+        Serializable fallbackValue = fallback.getValue();
+        return fallbackValue == null ? deprecatedValue : fallbackValue;
+    }
+
+    private StringBuilder newDeprecatedMessage() {
+        StringBuilder builder = new StringBuilder().append("Property '")
+                                                   .append(getPath())
+                                                   .append("' is marked as deprecated from '")
+                                                   .append(getSchema().getName())
+                                                   .append("' schema");
+        DeprecatedProperty deprecatedParent = getDeprecatedParent();
+        if (deprecatedParent != this) {
+            builder.append(" because property '")
+                   .append(deprecatedParent.getPath())
+                   .append("' is marked as deprecated");
+        }
+        return builder.append(", don't use it anymore.\n");
+    }
+
+    private DeprecatedProperty getDeprecatedParent() {
+        DeprecatedProperty property = this;
+        while (property.getParent() instanceof DeprecatedProperty) {
+            property = (DeprecatedProperty) property.getParent();
+        }
+        return property;
+    }
+
+    @Override
+    public boolean isContainer() {
+        return deprecated.isContainer();
+    }
+
+    @Override
+    public String getName() {
+        return deprecated.getName();
+    }
+
+    @Override
+    public Type getType() {
+        return deprecated.getType();
+    }
+
+    @Override
+    public Field getField() {
+        return deprecated.getField();
+    }
+
+    @Override
+    public Property get(String name) throws PropertyNotFoundException {
+        Property childFallback = fallback;
+        if (childFallback != null) {
+            childFallback = childFallback.get(name);
+        }
+        return new DeprecatedProperty(this, deprecated.get(name), childFallback);
+    }
+
+    @Override
+    public Property get(int index) throws PropertyNotFoundException {
+        // TODO wrap it in a DeprecatedProperty
```

Comment:
We can't add a NXP for every single TODO note in the code, they would accumulate as possible improvements and never be read anyway.